### ISSUE TITLE
IntegrationTestCase & SecurityComponent

### ISSUE THREAD

**Issue by steefaan** [cakephp/cakephp#7004]
It seems there isn't a nice way to simulate the `fields` and `unlocked` hashes of the SecurityComponent in a controller test case. The CSRF token is pretty easy to handle in an extended IntegrationTestCase and I wished it would be similar with the hashes for the SecurityComponent. Let me give you an example to show what I mean. This is a part of my extended IntegrationTestCase:

```
    protected function _buildRequest($url, $method, $data)
    {
        $csrfToken = Security::hash(Text::uuid(), 'sha1', true);
        $this->cookie('csrfToken', $csrfToken);

        if ($method !== 'GET') {
            $data['_csrfToken'] = $csrfToken;
        }

        return parent::_buildRequest($url, $method, $data);
    }
```

As you can see it's easy to handle the CSRF token here. Thanks to `$this->cookie()` at this point. Gladly I would do the same for the SecurityComponent but therefore I would need to copy and paste parts of the code from [FormHelper::secure()](https://github.com/cakephp/cakephp/blob/master/src/View/Helper/FormHelper.php#L551) to my extended IntegrationTestCase and this seems a bit messy.

**Comment by markstory** [comment]
If you could have a simpler to use API what would that look like?

**Comment by steefaan** [comment]
May this could be a solution:

```
trait SecurityTrait
{
    protected function _buildRequest($url, $method, $data)
    {
        if (in_array($method, ['POST', 'PUT'])) {
            $token = $this->_buildSecurityToken($url, array_keys($data));
            $data = Hash::merge($token, $data);
        }

        $csrfToken = Security::hash(Text::uuid(), 'sha1', true);
        $this->cookie('csrfToken', $csrfToken);

        if (in_array($method, ['PATCH', 'PUT', 'POST', 'DELETE'])) {
            $data['_csrfToken'] = $csrfToken;
        }

        return parent::_buildRequest($url, $method, $data);
    }

    protected function _buildSecurityToken($url, array $fields = [])
    {
        $locked = [];
        $tokens = ['_Token', '_csrfToken'];

        foreach ($fields as $key => $value) {
            if (is_numeric($value)) {
                $value = (string)$value;
            }
            if (in_array($value, $tokens)) {
                unset($fields['key']);
                continue;
            }
            if (!is_int($key)) {
                $locked[$key] = $value;
                unset($fields[$key]);
            }
        }

        sort($fields, SORT_STRING);
        ksort($locked, SORT_STRING);
        $fields += $locked;

        $locked = implode(array_keys($locked), '|');
        $hashParts = [$url, serialize($fields), '', Security::salt()];
        $fields = Security::hash(implode('', $hashParts), 'sha1');

        return [
            '_Token' => [
                'fields' => urlencode($fields . ':' . $locked),
                'unlocked' => ''
            ]
        ];
    }
}
```

**Comment by steefaan** [comment]
This way you have different possiblities. If you include the SecurityTrait to your controller test cases it will simulate a CSRF token and the Security token to prevent the black-holed error message. It's independ what data you will send with you `$this->post()` request in the test method, the token is always correct. You can event set your own 

```
_Token => [
    'fields' => 'your-token'
    'unlocked' => ''
]
```

to the post data if you want. Seems to be the most felxible approch for this general problem with test cases and SecurityComponent-

**Comment by burzum** [comment]
:+1:

**Comment by markstory** [comment]
A trait is an interesting approach, that would let us use the same code in the helper and the test cases.

**Comment by steefaan** [comment]
I don't want to rush you but are you able to give me estimate how long it will take to have it on the master branch? Maybe I can help you with something to speed the process?

**Comment by dereuromark** [comment]
A trait sounds great :+1:

**Comment by markstory** [comment]
@steefaan I don't have an estimate. Working in the evenings and weekends doesn't lend itself to giving reliable time estimates. If you need something fast, you might be best to use the solution you provided, or submit a pull request.

**Comment by markstory** [comment]
Fixed with #7524
